File: 974.subarray-sums-divisible-by-k.cpp
// @leet start

#include <unordered_map>
#include <vector>
using std::vector;

class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        // Concept similar to LC523.

        // TC O(n), SC O(k)
        std::unordered_map<int, int> map;
        int total = 0;
        int prefixSum = 0;

        ++map[0];

        for (const int& num : nums) {
            // Modulo because for 
            // prefixSum[i] % k == 4 and prefixSum[j] % k == 4, 
            // then the subarray between i and j fulfills condition
            prefixSum = (prefixSum + num) % k;

            if (prefixSum < 0) {
                // Deal with negative modulo issues
                prefixSum += k;
            }
            // No need to do [prefixSum - k] here because the % above does that
            total += map[prefixSum];
            ++map[prefixSum];
        }
        
        return total;
    }
};
// @leet end

File: 974.subarray-sums-divisible-by-k.py
# @leet imports start
from typing import List, Optional

# @leet imports end

from collections import defaultdict


# @leet start
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        counts = defaultdict(int)
        counts[0] = 1
        ans = 0
        psum = 0
        for x in nums:
            psum += x
            ans += counts[psum % k]
            counts[psum % k] += 1

        return ans


# @leet end

n = [4, 5, 0, -2, -3, 1]
k = 5
x = Solution().subarraysDivByK(n, k)
print(x)

s = [4, 9, 9, 7, 4, 5]
m = [4, 4, 4, 2, 4, 0]
a = [0, 1, 2, 0, 3, 0]

